### PR TITLE
[Mamba] Add FlashInfer ReplaySSM backend

### PR DESCRIPTION
**Pull request by askliar**
## Summary

This PR adds FlashInfer as an optional ReplaySSM backend for Mamba2. It calls FlashInfer's native `checkpointing_ssu` API, shares replay trackers per KV-cache group, and autotunes a maximum-batch decode through the existing FlashInfer warmup flow before CUDA-graph capture. FlashInfer ReplaySSM supports Model Runner V1 and V2; Triton ReplaySSM remains V1-only.

## Blocked on FlashInfer

> [!IMPORTANT]
> This PR is blocked by [flashinfer-ai/flashinfer#4545](https://github.com/flashinfer-ai/flashinfer/pull/4545), which adds native ReplaySSM autotuning. Do not merge this PR until #4545 is available in a compatible FlashInfer release and vLLM's package and Docker pins are updated together.

Older FlashInfer versions fail capability validation rather than silently using a non-native tuning path.

## Validation

Validation used `nvidia/NVIDIA-Nemotron-3.5-Lightning-30B-A3B-NVFP4` on H100 GPUs with MTP disabled.

### GSM8K

Matched 5-shot GSM8K over all 1,319 questions, temperature 0, and `max_num_seqs=64`:

| Backend | Correct | Accuracy | Invalid |
| --- | ---: | ---: | ---: |
| FlashInfer ReplaySSM | 1117/1319 | 84.685% | 0 |
| Triton ReplaySSM | 1119/1319 | 84.837% | 0 |

The paired exact test gives `p=0.913`, so the observed accuracy difference is not statistically significant.

### Throughput

Matched TP=1 fixed-length decode benchmark: 512 decode steps, 128 warmup steps, five measured repeats, and ReplaySSM buffer length 16.

| Batch | FlashInfer TPS | Triton TPS | FlashInfer delta |
| ---: | ---: | ---: | ---: |
| 1 | 366.22 | 360.17 | +1.68% |
| 16 | 2774.93 | 2901.00 | -4.35% |
| 64 | 6630.05 | 6725.68 | -1.42% |
| 128 | 9368.32 | 9452.60 | -0.89% |

FlashInfer is faster at batch size 1; Triton remains faster at the larger tested batch sizes.

Additional coverage: V1 and V2 smokes, two-node TP=2 and PP=2 smokes, focused H100 unit/integration tests

### CODE REVIEW

File: vllm/model_executor/layers/mamba/ops/ssu_dispatch.py [hunk 1]
```diff
@@ -3,25 +3,124 @@
...
+    prev = tl.load(prev_num_accepted + slots, mask=valid, other=0)
+    start = tl.load(ring_start + slots, mask=valid, other=0)
+    must_checkpoint = prev + 1 > logical_window
+    next_start = tl.where(
+        must_checkpoint,
+        (start + prev) % ring_buffer_len,
+        start,
+    )
+    next_prev = tl.where(must_checkpoint, 1, prev + 1)
+    tl.store(ring_start + slots, next_start, mask=valid)
```
[the remark this answers is not in the document]
Comment (askliar) [reply vllm/model_executor/layers/mamba/ops/ssu_dispatch.py]: Fixed by passing the tracker capacity to the Triton kernel and masking negative, null, and out-of-range slot IDs before every load/store. Added a CUDA regression covering `-1`, `num_states`, the null slot, and one valid slot.

**Comment by askliar** [comment]
Rebased onto current `main`; GitHub now reports the PR as mergeable. The cleanup also removes the generic block-table API change and `dummy_first_block_id`, simplifies the ReplaySSM helpers, and preserves the max-batch decode autotune path.

File: vllm/v1/worker/utils.py [hunk 2]
```diff
@@ -549,6 +553,44 @@ def bind_kv_cache(
...
+            )
+        mixer_groups.extend(
+            [mixer]
+            for layer_name, mixer in replayssm_mixers.items()
+            if layer_name not in grouped_names
+        )
+    else:
+        # Without cache groups, block-index namespaces cannot be proven equal.
+        mixer_groups = [[mixer] for mixer in replayssm_mixers.values()]
+    share_replayssm_ring_trackers(mixer_groups)
```
Comment (benchislett) [inline vllm/v1/worker/utils.py]: this seems a bit overengineered. why are we constructing this list just to pass it into a one-off helper function? 

seems much cleaner to just pass `ordered_layer_names` into the helper function and have it iterate+filter to only the MambaMixer2-relevant ones inline. The helper doesn't seem to be using the list for anything other than iteration, so it doesn't really need to be materialized.

File: vllm/v1/worker/gpu/model_runner.py [hunk 3]
```diff
@@ -1324,9 +1328,18 @@ def prepare_attn(
...
         block_tables = self.block_tables.get_dummy_block_tables(input_batch.num_reqs)
+        if valid_state_slots:
+            state_slots = torch.arange(
+                1,
+                input_batch.num_reqs + 1,
+                dtype=torch.int32,
+                device=self.device,
+            )
+            for block_table in block_tables:
+                block_table[:, 0].copy_(state_slots)
```
Comment (benchislett) [inline vllm/v1/worker/gpu/model_runner.py:1408]: what is this for?

File: vllm/v1/attention/backends/mamba_attn.py [hunk 4]
```diff
@@ -769,7 +804,7 @@ def _update_metadata_for_cudagraph_capture(
                     )
                     block_idx_last_scheduled_token_prev_step[metadata.num_decodes :] = 0
 
-            if self.use_replayssm:
+            if self.use_replayssm and not self.use_flashinfer_replayssm:
```
Comment (benchislett) [inline vllm/v1/attention/backends/mamba_attn.py:807]: to be clear, does this mean we're still supporting the Triton-backend replaySSM? Is this just for Kimi's RecoverSSM or will we also have dual support for Mamba2?

If so, we should check on it down the line and strip it away if flashinfer is serving our needs

File: vllm/model_executor/layers/mamba/mamba_mixer2.py [hunk 5]
```diff
@@ -1148,6 +1196,41 @@ def mamba_type(self) -> MambaAttentionBackendEnum:
...
+                )
+
+        ring_start = torch.zeros(expected[0], dtype=torch.int32, device=expected[1])
+        prev_num_accepted = torch.zeros_like(ring_start)
+        for mixer in mixers:
+            mixer._replayssm_ring_start = ring_start
+            mixer._replayssm_prev_num_accepted = prev_num_accepted
+            mixer._updates_replayssm_trackers = False
+
+        mixers[-1]._updates_replayssm_trackers = True
```
Comment (benchislett) [inline vllm/model_executor/layers/mamba/mamba_mixer2.py]: I don't see any clear assertion about proper ordering here. How do we know that the last one in the list is the "final local layer"? Where do the kv specs get ordered within the group?

File: vllm/model_executor/warmup/kernel_warmup.py [hunk 6]
```diff
@@ -234,6 +237,137 @@ def _flashinfer_autotune_skip_ops(runner: "GPUModelRunner") -> set[str] | None:
...
+def _replayssm_autotune_kwargs(
+    runner: "GPUModelRunner", max_token_prefill_kwargs: dict[str, Any]
+) -> tuple[int, dict[str, Any]] | None:
+    config = runner.vllm_config
+    if not (
+        config.cache_config.use_replayssm
+        and config.mamba_config.backend == MambaBackendEnum.FLASHINFER
+    ):
+        return None
+    use_v2_model_runner = config.use_v2_model_runner
```
Comment (benchislett) [inline vllm/model_executor/warmup/kernel_warmup.py]: nit: i think you can inline the use of this variable. factoring it out saves a few chars but adds lines, and is readability-negative IMO

File: vllm/model_executor/warmup/kernel_warmup.py [hunk 7]
```diff
@@ -234,6 +237,137 @@ def _flashinfer_autotune_skip_ops(runner: "GPUModelRunner") -> set[str] | None:
...
+        tensors = (
+            *module.kv_cache,
+            ring_start,
+            prev_num_accepted,
+        )
+        for tensor in tensors:
+            if tensor.numel():
+                reset_tensors.setdefault(tensor.data_ptr(), tensor)
+
+    use_v2_model_runner = runner.vllm_config.use_v2_model_runner
```
Comment (benchislett) [inline vllm/model_executor/warmup/kernel_warmup.py]: same nit as above

File: vllm/model_executor/warmup/kernel_warmup.py [hunk 8]
```diff
@@ -234,6 +237,137 @@ def _flashinfer_autotune_skip_ops(runner: "GPUModelRunner") -> set[str] | None:
...
+        if use_v2_model_runner
+        else runner.uniform_decode_query_len
+    )
+    max_num_reqs = min(
+        runner.scheduler_config.max_num_seqs,
+        runner.max_num_tokens // query_len,
+        runner.kv_cache_config.num_blocks - 1,
+    )
+    if max_num_reqs <= 0:
+        logger.warning_once(
```
Comment (benchislett) [inline vllm/model_executor/warmup/kernel_warmup.py]: What does this mean? Is this reachable?

File: tests/test_config.py [hunk 9]
```diff
@@ -112,6 +112,41 @@ def test_v2_model_runner_env_tri_state(monkeypatch, env_value, expected):
...
+        ),
+        model_config=None,
+        num_speculative_tokens=0,
+        mamba_config=SimpleNamespace(backend=backend),
+        use_v2_model_runner=use_v2_model_runner,
+        kv_transfer_config=None,
+    )
+
+
+def test_v2_replayssm_requires_flashinfer():
```
Comment (benchislett) [inline tests/test_config.py]: are these tests necessary? do you foresee a regression here?

**Review by benchislett — CHANGES_REQUESTED** [review]
Overall, I am comfortable with this integration, with the following caveats:
- Correctness is heavily reliant on the FlashInfer kernels and we should thoroughly validate the functionality here. We will probably want at least one end-to-end serving test hooked up to CI, ideally using ModelRunnerV2. I would also like to see more comprehensive correctness validation: run on Blackwell (if kernel supports), run with TP > 1, try with EP and/or DP (if supported), and please run something like MT-Bench and double-check the model outputs are sane under temperature=1 + top-p 0.95
- The unit tests seem a bit too comprehensive. Please do a review pass over these and prune out the ones that you think are not needed to carry forward maintaining. 
- I am of course not an expert in the low-level kernel integrations here. w.r.t. precision and algorithmic correctness I defer to Harris and Amey. AFAIK, this is a straightforward integration with no clear downsides.
- I left a couple nits / clarifying questions which I would like to see resolved.

File: vllm/v1/worker/gpu/model_runner.py (hunk 3 above)
Comment (askliar) [reply vllm/v1/worker/gpu/model_runner.py:1408]: Purely for autotuning of ReplaySSM. This is trying to replicate running multiple parallel requests for ReplaySSM.

File: vllm/v1/attention/backends/mamba_attn.py (hunk 4 above)
Comment (askliar) [reply vllm/v1/attention/backends/mamba_attn.py:807]: This means that we support both Triton and FlashInfer backend for ReplaySSM, yes. I think it's a better idea to support both, we can discuss it offline if required.

RecoverSSM is important for MTP PR, but not for this one.

File: vllm/v1/worker/utils.py (hunk 2 above)
Comment (askliar) [reply vllm/v1/worker/utils.py]: Fair enough, will fix!

File: vllm/model_executor/layers/mamba/mamba_mixer2.py (hunk 5 above)
Comment (askliar) [reply vllm/model_executor/layers/mamba/mamba_mixer2.py]: Will make it more explicit + use assert

File: vllm/model_executor/warmup/kernel_warmup.py (hunk 8 above)
Comment (askliar) [reply vllm/model_executor/warmup/kernel_warmup.py]: Removed.

**Comment by askliar** [comment]
@benchislett I've addressed most of the comments - please, let me know if the latest version looks good to you.

Main test runs I've completed are (all with N3.5 Lightning NVFP4): 
1. GB200/H100 TPS runs for regular single-GPU.
2. GB200/H100 GSM8k validations across different partitioning schemas.

For 1:
### H100

| Baseline TPS | FI ReplaySSM TPS | Triton ReplaySSM TPS |
|---:|---:|---:|
| 390.50 | 387.69 (-0.72% vs baseline; +1.95% vs Triton) | 380.27 (-3.31% vs baseline; -1.91% vs FI) |
| 2,922.48 | 2,940.48 (+0.62% vs baseline; -3.92% vs Triton) | 3,060.55 (+6.46% vs baseline; +4.08% vs FI) |
| 6,063.88 | 6,575.50 (+8.44% vs baseline; -0.80% vs Triton) | 6,628.40 (+8.68% vs baseline; +0.80% vs FI) |
| 8,203.61 | 9,242.77 (+12.67% vs baseline; -0.45% vs Triton) | 9,284.60 (+12.36% vs baseline; +0.45% vs FI) |
| 11,513.71 | 13,591.81 (+18.05% vs baseline; -0.99% vs Triton) | 13,727.26 (+19.11% vs baseline; +1.00% vs FI) |
| 13,329.24 | 16,434.66 (+23.30% vs baseline; +0.22% vs Triton) | 16,398.29 (+23.51% vs baseline; -0.22% vs FI) |

### GB200

| Baseline TPS | FI ReplaySSM TPS | Triton ReplaySSM TPS |
|---:|---:|---:|
| 439.27 | 438.16 (-0.25% vs baseline; +5.65% vs Triton) | 414.70 (-7.34% vs baseline; -5.35% vs FI) |
| 3,687.06 | 3,823.27 (+3.69% vs baseline; -0.19% vs Triton) | 3,830.55 (+0.86% vs baseline; +0.19% vs FI) |
| 8,325.10 | 8,743.25 (+5.02% vs baseline; +2.64% vs Triton) | 8,518.40 (+1.53% vs baseline; -2.57% vs FI) |
| 11,963.44 | 12,892.34 (+7.76% vs baseline; +3.42% vs Triton) | 12,466.08 (+3.85% vs baseline; -3.31% vs FI) |
| 17,657.23 | 19,492.70 (+10.39% vs baseline; +3.02% vs Triton) | 18,922.15 (+6.50% vs baseline; -2.93% vs FI) |
| 21,008.20 | 23,847.05 (+13.51% vs baseline; +3.35% vs Triton) | 23,073.70 (+9.38% vs baseline; -3.24% vs FI) |

For 2:
### Distributed GSM8K

| Hardware | Topology | No ReplaySSM | FI ReplaySSM | Delta |
|---|---|---:|---:|---:|
| H100 | TP2 | 84.230% | 84.003% | -0.227 pp |
| H100 | DP2 | 84.761% | 84.382% | -0.379 pp |
| H100 | DP2+EP | 84.685% | 84.079% | -0.607 pp |
| GB200 | TP2  | 85.216% | 84.610% | -0.607 pp |
| GB200 | DP2 | 84.230% | 84.534% | +0.303 pp |
| GB200 | DP2+EP | 84.382% | 84.079% | -0.303 pp |

Note that these differences are within noise range - I did not see any consistent degradation across runs.

File: vllm/model_executor/layers/mamba/ops/ssu_dispatch.py [hunk 10]
```diff
@@ -262,15 +358,113 @@ def __call__(
...
+
+
+@cache
+def flashinfer_replayssm_autotune_supported() -> bool:
+    """Return True when FlashInfer exposes ReplaySSM autotuning."""
+    try:
+        from flashinfer.mamba.checkpointing_ssu import CheckpointingSSURunner
+    except ImportError:
+        return False
+    return callable(CheckpointingSSURunner)
```
Comment (benchislett) [inline vllm/model_executor/layers/mamba/ops/ssu_dispatch.py]: Why `callable(...)` instead of just `is not None` or `True`?

Comment (askliar) [reply vllm/model_executor/layers/mamba/ops/ssu_dispatch.py]: Was completely unnecessary. Removed.

File: vllm/model_executor/warmup/kernel_warmup.py [hunk 11]
```diff
@@ -272,6 +278,135 @@ def _run_flashinfer_autotune_dummy_runs(runner: "GPUModelRunner") -> None:
         )
 
 
+def _replayssm_autotune_kwargs(
```
Comment (benchislett) [inline vllm/model_executor/warmup/kernel_warmup.py]: The replaySSM autotuning code can be moved into a separate file, following the others:
```
from vllm.model_executor.layers.mamba.ops.ssu_dispatch import (
    flashinfer_replayssm_autotune_supported,
)
from vllm.model_executor.warmup.b12x_warmup import b12x_warmup
from vllm.model_executor.warmup.cutedsl_warmup import cutedsl_warmup
from vllm.model_executor.warmup.deep_gemm_warmup import deep_gemm_warmup
from vllm.model_executor.warmup.deepseek_v4_mhc_warmup import (
    deepseek_v4_mhc_warmup,
)
from vllm.model_executor.warmup.fa4_cutedsl_warmup import (
    fa4_cutedsl_warmup,
)
```

If possible, we should keep replayssm-specific logic self-contained, and keep existing core functions (like `flashinfer_autotune`) minimally affected.

File: vllm/model_executor/warmup/kernel_warmup.py [hunk 12]
```diff
@@ -332,6 +474,10 @@ def flashinfer_autotune(runner: "GPUModelRunner") -> None:
             fi_utils.autotune(tune_mode=True, **autotune_kwargs),
         ):
             _run_flashinfer_autotune_dummy_runs(runner)
+            if replayssm_autotune is not None:
+                max_num_reqs, max_batch_decode_kwargs = replayssm_autotune
+                with _temporary_replayssm_autotune_state(runner, max_num_reqs):
+                    runner._dummy_run(**max_batch_decode_kwargs)
```
Comment (benchislett) [inline vllm/model_executor/warmup/kernel_warmup.py]: This is mixing abstraction levels a bit: put it in a helper (in a separate file: see above comment) and call it directly, next to `_run_flashinfer_autotune_dummy_runs`.

**Comment by askliar** [comment]
Addressed the comment here: https://github.com/vllm-project/vllm/pull/52928#issuecomment-5488777083

**Review by benchislett — APPROVED** [review]
